### PR TITLE
WIP: pull "kubectl edit" from openshift upstream

### LINKED ISSUE #8263: Upstream Openshift edit functionality into kubectl
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/edit.go

### PR DESCRIPTION
**Pull request by caesarxuchao**
pull [kubectl edit](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/edit.go) from openshift upstream.
add a test for it in test-cmd.sh

fix #8263 

@bgrant0607 @smarterclayton

### CODE REVIEW

**Comment by a-robinson** [comment]
cc @jlowdermilk

**Comment by smarterclayton** [comment]
@deads2k can you review since you reviewed edit before?

On Jun 29, 2015, at 6:30 PM, Kubernetes Bot <notifications@github.com>
wrote:

GCE e2e build/test *passed* for commit 1f15c5e
<https://github.com/GoogleCloudPlatform/kubernetes/commit/1f15c5e389b8d87a8d4131cda8e6e5ffbd7216dc>
.

   - Build Log
   <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/1f15c5e389b8d87a8d4131cda8e6e5ffbd7216dc/kubernetes-pull-build-test-e2e-gce/1269/build-log.txt>
   - Test Artifacts
   <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/1f15c5e389b8d87a8d4131cda8e6e5ffbd7216dc/kubernetes-pull-build-test-e2e-gce/1269/_artifacts/>
   - Internal Jenkins Results
   <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//1269>

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/10502#issuecomment-116865760>
.

File: hack/test-cmd.sh [hunk 1]
```diff
@@ -393,6 +393,15 @@ for version in "${kube_api_versions[@]}"; do
   # Post-condition: spec.container.name = "update-k8s-serve-hostname"
   kube::test::get_object_assert 'pod valid-pod' "{{(index .spec.containers 0).name}}" 'update-k8s-serve-hostname'
 
+  ## kubectl edit can update the image field of a POD. tmp-editor.sh is a fake editor
+  echo -e '#!/bin/bash\nsed -i "s/nginx/gcr.io\/google_containers\/serve_hostname/g" $1' > tmp-editor.sh
```
Comment (deads2k) [inline hack/test-cmd.sh]: What is `cwd` here? Something under `/tmp`?

File: pkg/kubectl/cmd/edit.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util/editor"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util/jsonmerge"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	//	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
```
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: remove cruft

File: pkg/kubectl/cmd/edit.go [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+  // Use an alternative editor
+  $ OSC_EDITOR="nano" kubectl edit dc/my-deployment
+
+  // Edit the service 'docker-registry' in JSON using the v1beta3 API format:
+  $ kubectl edit svc/docker-registry --output-version=v1beta3 -o json`
+)
+
+var errExit = fmt.Errorf("exit directly")
+
+func NewCmdEdit(f *cmdutil.Factory, out io.Writer) *cobra.Command {
```
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: Any chance of moving our `name`, and `fullName` pattern up here?  It makes it easier to compose commands together.

File: pkg/kubectl/cmd/edit.go [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+		},
+	}
+	usage := "Filename, directory, or URL to file to use to edit the resource"
+	kubectl.AddJsonFilenameFlag(cmd, &filenames, usage)
+	cmd.Flags().StringP("output", "o", "yaml", "Output format. One of: yaml|json.")
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	return cmd
+}
+
+func RunEdit(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, filenames util.StringList) error {
```
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: Since I get a second shot, I prefer distilling the `cmd` and `factory` down to the required options separately and then having the main `Run` be divorced from how it was represented on the command line.  Doing it like this means that you have to keep track of what the options were called throughout a relatively large method.

File: pkg/kubectl/cmd/edit.go [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+			return preservedFile(err, file, cmd.Out())
+		}
+
+		// attempt to calculate a delta for merging conflicts
+		delta, err := jsonmerge.NewDelta(original, edited)
+		if err != nil {
+			glog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
+			delta = nil
+		} else {
+			delta.AddPreconditions(jsonmerge.RequireKeyUnchanged("apiVersion"))
```
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: Probably want to add `kind` and if present `metadata.name`, `metadata.resourceversion`, `metadata.uid`, and `metadata.namespace`.

**Comment by deads2k** [comment]
Looks like a pretty straight port.  Should probably add a test for `RequireKeyUnchanged`.  I feel like that broke on us at some point during the release.

File: pkg/kubectl/cmd/edit.go (hunk 3 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: Thank you @deads2k. Is this specific to the "edit" command? If not, I'd like to keep it the same as other commands in k8s in this PR, and send another PR to refactor all commands.

File: pkg/kubectl/cmd/edit.go [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+	//	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	//	"github.com/openshift/origin/pkg/cmd/util/editor"
+	//"github.com/openshift/origin/pkg/util/jsonmerge"
+)
+
+const (
+	editLong = `Edit a resource from the default editor.
+
+The edit command allows you to directly edit any API resource you can retrieve via the
+command line tools. It will open the editor defined by your OSC_EDITOR, GIT_EDITOR,
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/edit.go]: We shouldn't use OSC_EDITOR -- that's openshift. We should use KUBE_EDITOR.

File: pkg/kubectl/cmd/edit.go [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+  $ kubectl edit svc/docker-registry
+
+  // Edit the DeploymentConfig named 'my-deployment':
+  $ kubectl edit dc/my-deployment
+
+  // Use an alternative editor
+  $ OSC_EDITOR="nano" kubectl edit dc/my-deployment
+
+  // Edit the service 'docker-registry' in JSON using the v1beta3 API format:
+  $ kubectl edit svc/docker-registry --output-version=v1beta3 -o json`
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/edit.go]: Should be v1 example.

File: pkg/kubectl/cmd/edit.go [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+
+		visitor := resource.NewFlattenListVisitor(updates, rmap)
+
+		// need to make sure the original namespace wasn't changed while editing
+		if err = visitor.Visit(resource.RequireNamespace(cmdNamespace)); err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+
+		// attempt to calculate a delta for merging conflicts
+		delta, err := jsonmerge.NewDelta(original, edited)
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/edit.go]: Is this using strategic merge?

File: pkg/kubectl/cmd/edit.go [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+			return preservedFile(err, results.file, cmd.Out())
+		}
+		if err := printer.PrintObj(obj, buf); err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+		original := buf.Bytes()
+
+		// launch the editor
+		edit := editor.NewDefaultEditor()
+		edited, file, err := edit.LaunchTempFile("oc-edit-", ext, buf)
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/edit.go]: kubectl-edit-

File: pkg/kubectl/cmd/edit.go (hunk 8 above)
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: > Is this using strategic merge?

Strategic merge?

File: pkg/kubectl/cmd/edit.go [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+}
+
+type patchError struct {
+	error
+}
+
+// applyPatch reads the latest version of the object, writes it to version, then attempts to merge
+// the changes onto it without conflict. If a conflict occurs jsonmerge.IsConflicting(err) is
+// true. The info object is mutated
+func applyPatch(delta *jsonmerge.Delta, info *resource.Info, version string) error {
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/edit.go]: Is this using strategic merge?

Declarative update ("apply") will probably be similar. cc @jackgr @ghodss

File: pkg/kubectl/cmd/edit.go (hunk 4 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @deads2k, I'm not sure if I follow you. Do you mean instead of passing in `cmd` and `factory`, we should pass in things like `cmdutil.GetFlagString(cmd, "output")`?

Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: > I'm not sure if I follow you. Do you mean instead of passing in cmd and factory, we should pass in things like cmdutil.GetFlagString(cmd, "output")

Yes, it contains the knowledge of flags outside of the main logic and that makes unit testing and command delegation problems easier.  When I started the review, I didn't realize it wasn't @smarterclayton's pull.  I don't mind delaying this until we try to rebase back downstream.

**Comment by bgrant0607** [comment]
cc @mikedanese

File: pkg/kubectl/cmd/edit.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/edit.go]: We should probably nuke KUBE_EDITOR and tell people to set EDITOR.  I only
copied that pattern from GIT, but in practice anyone who wants it can just
set it on their command line.

On Tue, Jun 30, 2015 at 12:51 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/edit.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10502#discussion_r33595301>
> :
>
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util/editor"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util/jsonmerge"
> > +
> > +	"github.com/golang/glog"
> > +	"github.com/spf13/cobra"
> > +
> > +	//	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
> > +	//	"github.com/openshift/origin/pkg/cmd/util/editor"
> > +	//"github.com/openshift/origin/pkg/util/jsonmerge"
> > +)
> > +
> > +const (
> > +	editLong = `Edit a resource from the default editor.
> > +
> > +The edit command allows you to directly edit any API resource you can retrieve via the
> > +command line tools. It will open the editor defined by your OSC_EDITOR, GIT_EDITOR,
>
> We shouldn't use OSC_EDITOR -- that's openshift. We should use KUBE_EDITOR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10502/files#r33595301>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by bgrant0607** [comment]
See also #8662.

File: pkg/kubectl/cmd/edit.go (hunk 8 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: No, it's used as a merge patch (http://tools.ietf.org/html/draft-ietf-appsawg-json-merge-patch-07)

FWIW, different patch operations are described here (https://github.com/GoogleCloudPlatform/kubernetes/blob/5520386b180d3ddc4fa7b7dfe6f52642cc0c25f3/docs/api-conventions.md#patch-operations)

File: pkg/kubectl/cmd/edit.go [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+	rmap := &resource.Mapper{
+		ObjectTyper:  typer,
+		RESTMapper:   mapper,
+		ClientMapper: f.ClientMapperForCommand(),
+	}
+
+	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(filenames...).
+		//SelectorParam(selector).
```
Comment (mikedanese) [inline pkg/kubectl/cmd/edit.go]: remove commented out code.

File: pkg/kubectl/cmd/edit.go [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+		ResourceTypeOrNameArgs(true, args...).
+		Latest()
+	if err != nil {
```
Comment (mikedanese) [inline pkg/kubectl/cmd/edit.go]: This is the err from f.DefaultNamespace() which you have already checked. I think you want s/err/b.Err()/

File: pkg/kubectl/cmd/edit.go [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+		r.header.reasons = append(r.header.reasons, reason)
+		return fmt.Sprintf("Error: the %s %s is invalid", info.Mapping.Kind, info.Name)
+	case errors.IsNotFound(err):
+		r.notfound++
+		return fmt.Sprintf("Error: the %s %s has been deleted on the server", info.Mapping.Kind, info.Name)
+
+	case errors.IsConflict(err):
+		if r.delta != nil {
+			v1 := info.ResourceVersion
+			if perr := applyPatch(r.delta, info, r.version); perr != nil {
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @deads2k, I actually don't understand the logic here. If Update() returns a conflict, why bother trying to make an update in applyPatch() again? Is it hoping that the jsonmerge package will find out there is actually no conflict?

File: pkg/kubectl/cmd/edit.go [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+	if err != nil {
+		return err
+	}
+
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	r := b.Flatten().Do()
```
Comment (mikedanese) [inline pkg/kubectl/cmd/edit.go]: Since you don't make any other modifications to the builder, instead of doing this here, you should chain this on the resource builder, then immediatly check if there was an err from Do(). Then you need to check r.Err() to make sure there was no err.

File: pkg/kubectl/cmd/edit.go (hunk 13 above)
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: > @deads2k, I actually don't understand the logic here. If Update() returns a conflict, why bother trying to make an update in applyPatch() again? Is it hoping that the jsonmerge package will find out there is actually no conflict?

The conflict is based on update detecting that the `resourceVersion` does not match.  That happens very commonly when a controller is updating `status`.  In such a case, you can diff the `editedObject` against the `originalObject` and the `currentObject` against the `originalObject` and compare the diffs.  If the diffs don't conflict, then you can apply the `editDiff` against the `currentObject` and issue an update.  That preserves user intent.

Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: I see. Thank you for the explanation.

File: pkg/kubectl/cmd/edit.go (hunk 6 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: I think it's a good idea to use the existing env variable. I'll take @smarterclayton 's suggestion.

File: pkg/kubectl/cmd/edit.go (hunk 12 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: nice catch!

File: pkg/kubectl/cmd/edit.go (hunk 10 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: no, merge patch.

**Comment by mikedanese** [comment]
@deads2k @smarterclayton why is edit even doing a merge. It seems like this should be a replace as the merge is done manually. Is that just for resource version conflicts?

File: pkg/kubectl/cmd/edit.go [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+	}
+	data, err := info.Mapping.Codec.Encode(obj)
+	if err != nil {
+		return patchError{err}
+	}
+	merged, err := delta.Apply(data)
+	if err != nil {
+		return patchError{err}
+	}
+	updated, err := resource.NewHelper(info.Client, info.Mapping).Update(info.Namespace, info.Name, false, merged)
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @mikedanese points out maybe the user don't want to send the merged data, rather, if there is no conflict, they want to send their edits as they are, and ignore any changes made by other parties. @bgrant0607 @smarterclayton @deads2k, what do you think?

Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: > @mikedanese points out maybe the user don't want to send the merged data, rather, if there is no conflict, they want to send their edits as they are, and ignore any changes made by other parties. @bgrant0607 @smarterclayton @deads2k, what do you think?

What we found was that it was really hard to use `edit` on resources that had `status` getting updated often. For instance, nodes are updated all the time so it was almost impossible to hit the `edit` window on them.  Having two non-conflicting updates applied via a diff seems more consistent with other deconflicting approaches.  Git doesn't force manual merging for two non=conflicting updates in a single file and it works pretty well.

Comment (mikedanese) [inline pkg/kubectl/cmd/edit.go]: Okay, that makes sense. That's an argument for splitting spec and status into separate etcd keys and making the pod a virtual resource.

Okay since we have to patch: We also have extensive server side code to apply multiple types of patches (including merge-patch). I'm not opposed to generating diffs in the client but I think we should be sending diffs as a PATCH, not applying them and sending the result as a PUT.

Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @deads2k, what we are suggesting is when there is no conflict, we just update the resource version of user's edits, and send it to Update(). This shouldn't make `edit` any more difficult.

Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: I think I prefer @mikedanese's idea of generating diffs and sending a patch (https://github.com/GoogleCloudPlatform/kubernetes/pull/10502/files#r33606922)

Comment (mikedanese) [inline pkg/kubectl/cmd/edit.go]: @caesarxuchao as we discussed the two approaches (or three) should be essentially the same in practice. Let's choose the one that we can do in the least lines of code, and that makes our client the skinniest.

File: pkg/kubectl/cmd/edit.go (hunk 5 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @deads2k, because we are going to use Patch(), instead of Update(), I think we can rely on the server to do these checks.

Comment (smarterclayton) [inline pkg/kubectl/cmd/edit.go]: Kind actually is somewhat dangerous to change, as is name, because they will not convey the users original intent (edit resource X of type Y) to the server.  The server would give a message like "Y is not a valid kind" or "X does not exist", not the more appropriate "you cannot rename X".

We should at minimum check "kind" - name may not exist on all API objects you edit.

Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: Agree.

File: pkg/kubectl/cmd/edit.go [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+			results.delta = delta
+			results.version = defaultVersion
+		}
+
+		err = visitor.Visit(func(info *resource.Info) error {
+			data, err := info.Mapping.Codec.Encode(info.Object)
+			if err != nil {
+				return err
+			}
+			updated, err := resource.NewHelper(info.Client, info.Mapping).Update(info.Namespace, info.Name, false, data)
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @smarterclayton @deads2k it seems the preconditions are not checked before this Update(), so it's possible for the user to change kind/apiversion/name completely, i.e. a user can start `kubectl edit` with one object, and end up update another object. Please correct me if I'm wrong.

Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: I just confirmed that in the current version, I can call `kubectl edit pod pod1`, and copy paste the output of `kubectl get pod pod2` into the editor, and update pod2.

For a single object, it's easy to do a check before calling Update() here. For a List, it will be harder, we need to check the kind/name fields of all enclosed objects.

File: pkg/kubectl/cmd/edit.go (hunk 13 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/edit.go]: @deads2k, this applyPatch() approach doesn't work for a list of objects, because `r.delta` is for the list and `info` is for individual object.

File: pkg/kubectl/cmd/util/jsonmerge/jsonmerge.go [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+	return err == ErrPreconditionFailed
+}
+
+var ErrPreconditionFailed = fmt.Errorf("a precondition failed")
+var ErrConflict = fmt.Errorf("changes are in conflict")
+
+// hasConflicts returns true if the left and right JSON interface objects overlap with
+// different values in any key.  The code will panic if an unrecognized type is passed
+// (anything not returned by a JSON decode).  All keys are required to be strings.
+func hasConflicts(left, right interface{}) bool {
```
Comment (caesarxuchao) [inline pkg/kubectl/cmd/util/jsonmerge/jsonmerge.go]: @deads2k, it returns true when right side is map[] (i.e. no change on the server), I think this is not the intended behavior.

File: pkg/kubectl/cmd/edit.go (hunk 8 above)
Comment (bgrant0607) [inline pkg/kubectl/cmd/edit.go]: Specifically: https://github.com/GoogleCloudPlatform/kubernetes/blob/ed985b86544f4b129c724d160f9c300e9203eb1a/docs/api-conventions.md#strategic-merge-patch

This is what we use for `kubectl patch` and is what we plan to use for `kubectl apply`.

The key difference is that maps encoded as lists are merged like maps, by merging elements with matching keys.

**Comment by bgrant0607** [comment]
@caesarxuchao Thanks for trying, but it looks like a number of non-trivial changes are required. Let's backburner this until after 1.0.

**Comment by bgrant0607** [comment]
cc @jlowdermilk

**Comment by jackgr** [comment]
I'm looking at diff+patch again after a long reprieve. What is the current status of the work on this PR, especially [the 3 way diff](https://github.com/GoogleCloudPlatform/kubernetes/issues/1702#issuecomment-115792182) from @smarterclayton?

**Comment by ghodss** [comment]
@jackgr There has been a lot of activity on #8662, but I don't think work on proper three-way merge has started (though it needs to be done).

**Comment by bgrant0607** [comment]
Correct, no 3-way merge has been started, AFAIK. Would love to see progress on that.

**Comment by caesarxuchao** [comment]
cc @JanetKuo
